new-new-packages/compiler/src/compiler/string_to_rcst.rs

```diff
@@ -0,0 +1,1394 @@
+        let mut has_weird_whitespace = false;
+        let mut indent_in_spaces = 0;
+
+        while indent_in_spaces < 2 * indentation {
+            let c = input.chars().next()?;
+            let (is_weird, indent_bonus) = match c {
+                ' ' => (false, 1),
+                '\t' => (true, 2),
+                c if c.is_whitespace() => (true, 1),
+                _ => return None,
+            };
+            chars.push(c);
+            has_weird_whitespace |= is_weird;
+            indent_in_spaces += indent_bonus;
+            input = &input[c.len_utf8()..];
+        }
+        let whitespace = chars.into_iter().join("");
+        Some(if has_weird_whitespace {
+            (
+                input,
+                Rcst::Error {
+                    unparsable_input: whitespace,
+                    error: RcstError::WeirdWhitespaceInIndentation,
+                },
+            )
+        } else {
+            (input, Rcst::Whitespace(whitespace))
+        })
+    }
+    #[test]
+    fn test_leading_indentation() {
+        assert_eq!(
+            leading_indentation("foo", 0),
+            Some(("foo", Rcst::Whitespace("".to_string())))
+        );
+        assert_eq!(
+            leading_indentation("  foo", 1),
+            Some(("foo", Rcst::Whitespace("  ".to_string())))
+        );
+        assert_eq!(leading_indentation("  foo", 2), None);
+    }
+
+    /// Consumes all leading whitespace (including newlines) and comments that
+    /// are still within the given indentation. Won't consume newlines before a
+    /// lower.
+    pub fn whitespaces_and_newlines(
+        mut input: &str,
+        indentation: usize,
+        also_comments: bool,
+    ) -> (&str, Vec<Rcst>) {
+        log::trace!(
+            "whitespaces_and_newlines({:?}, {:?}, {:?})",
+            input,
+            indentation,
+            also_comments
+        );
+        let mut parts = vec![];
+        if let Some((new_input, whitespace)) = single_line_whitespace(input) {
+            input = new_input;
+            parts.push(whitespace);
+        }
+
+        let mut input = input;
+        loop {
+            if also_comments {
+                if let Some((i, whitespace)) = comment(input) {
+                    input = i;
+                    parts.push(whitespace);
+                }
+            }
+
+            // We only consume newlines if there is sufficient indentation
+            // coming after.
+            let mut new_input = input;
+            let mut new_parts = vec![];
+            while let Some((new_new_input, newline)) = newline(new_input) {
+                new_parts.push(newline);
+                new_input = new_new_input;
+            }
+            if new_input == input {
+                break; // No newlines.
+            }
+            match leading_indentation(new_input, indentation) {
+                Some((new_input, whitespace)) => {
+                    new_parts.push(Rcst::Whitespace(whitespace.to_string()));
```

Comment:
`whitespace` should already be an `Rcst::Whitespace`

```suggestion
                    new_parts.push(whitespace);
```